### DISCUSSION TITLE
V10.0.0 Breaking changes

### DISCUSSION THREAD

**Discussion by andyblac** [BottlecapDave/HomeAssistant-OctopusEnergy#678] [General]
> BREAKING CHANGES
> The current and previous accumulative gas consumptions meter entity ids have been updated to include suffixes representing the reported unit of measurement. Any references to these entities will need to be updated.

does this if I'm using `sensor.octopus_energy_gas_xxxxx_current_accumulative_consumption`, in my my Energy Dashboard, I will loose my history after updating to the new entity name ?

I hope not, else I will be stuck on the old version.

**Comment by BottlecapDave** [comment]
~~This was the case unfortunately for v10.0.0, however I believe I've managed to find an undocumented API that will support the migration. The breaking change will still be applicable as the names will still change, but the history should come along for the ride. For anyone coming to this thread, this means you might need to do some additional work to see the history.~~

~~If you are on a version **below** v10.0.0, then you should upgrade straight to v10.0.1 (or above). This will mean the migration process will take effect and you shouldn't need to do anything~~

~~If you are on version v10.0.0 and migrate to v10.0.1, you will likely see some warnings stating that the migration failed. This is because you have two entities present, one with the old name and old historic data and one with the new name and no history. In this scenario you have two options~~
1. ~~Delete the old entity and live with the historic data being gone. If you restart HA the warning should be gone~~
2. ~~Downgrade to a version **before** v10.0.0 and delete the unavailable entities that were created as part of v10.0.0 (`sensor.octopus_energy_gas_{{METER_SERIAL_NUMBER}}_{{MPRN_NUMBER}}_previous_accumulative_consumption_m3` and `sensor.octopus_energy_gas_{{METER_SERIAL_NUMBER}}_{{MPRN_NUMBER}}_current_accumulative_consumption_kwh`). If you have already deleted the old entities you should be able to get them back if you restore from a backup. Once these entities have been removed, you should then upgrade to v10.0.1 or above and restart HA which should update the names of the entities and bring the historic data with it.~~

Issues have been discovered with the migration logic. This is still being investigated if it's actually possible as the HA APIs that are available don't seem to be migrating as intended. A new update with this migration logic removed has been released and v10.0.1 has been removed to avoid any further damage.

**Comment by andyblac** [comment]
thanks for your work, I will try upgrading to v10.0,1 tomorrow (ill take backup just before just incase), and report back how it goes.

**Comment by ksimuk** [comment]
Ahh 😫, the second breaking change in the row, I've renamed the sensor to keep the history, and now the breaking change again.

**Reply by BottlecapDave** [reply]
Sorry, I didn't consider this a breaking change from v10 as everything still worked as described in the last update (you just got a warning) and you shouldn't have gotten the warning if you'd already renamed the entity and had removed the old one. However in my testing, renaming the entity alone wasn't enough to keep the history.

**Reply by andyblac** [reply]
how did you get it to work, can you provide a step by  step  please ?

did you rename the old entity before updating ?

**Reply by BottlecapDave** [reply]
I did exactly as described above. If you were on a version pre v10, you shouldn't have needed to do anything as the entity name should have updated itself

**Reply by andyblac** [reply]
I'm on v9, did the update and restarted, energy dashboard said entity was missing, I change to new name, and lost all gas history..

v9 entity name is `sensor.octopus_energy_gas_xxxxx_xxxxx_current_accumulative_consumption` that is used in dashboard. I change it to `sensor.octopus_energy_gas_xxxx_xxxxx_current_accumulative_consumption_kwh`

**Reply by BottlecapDave** [reply]
I'm reverting my test box to try and see what could have happened. Unfortunately I haven't done a full back up in a while so need to wait for some data to be populated on the entities in question before doing a backup and then trying the upgrade process.

**Reply by ksimuk** [reply]
I saw braking change notice, realised entity was renamed, updated the integration, removed old entity and renamed new to the old name immediately after, for some reason it worked for me and I have the history, but it is using old name without `_kwh`. Updated to v10.0.2 now avoiding v10.0.1.

**Reply by BottlecapDave** [reply]
That's interesting I thought the deletion of entities results in the stats being removed. Using the old name is fine as long as you remember when things are referencing the new name :) Thanks for sharing your experience.

**Reply by alfwro13** [reply]
I have used that method in the past and now and it does work - https://community.home-assistant.io/t/how-to-replace-entity-in-energy-dashboard-by-new-one-without-losing-history/545933/2

**Reply by andyblac** [reply]
and the Octopus integration will use the old entity ok ? or will it look for the new one ? `xxx_kwh`

**Reply by alfwro13** [reply]
yep, as far I know you can rename all your sensors regardless of what integration provides them to whatever name you want.

**Reply by BottlecapDave** [reply]
If this now works (I don't remember it working when I made the last name change), then that would be ideal and I can focus my available time on other issues/features.

**Reply by andyblac** [reply]
It seems to have worked, still got my history data, and new data is updating.

**Reply by BottlecapDave** [reply]
This is excellent to hear. It sounds like this is the safest migration solution, so I'll park any automated approach which might produce undesired effects (the solution in 10.0.1 should have gone through the same flow as editing via the UI but something obviously went wrong). 

I try and keep name changes to a minimum, but the inconsistency got me this time :( I'm sorry for the inconvience.

**Reply by andyblac** [reply]
btw I had to do the same with `sensor.octopus_energy_gas_xxx_xxx_previous_accumulative_consumption_m3`, delete,  `sensor.octopus_energy_gas_xxx_xxx_previous_accumulative_consumption` and rename `sensor.octopus_energy_gas_xxx_xxx_previous_accumulative_consumption_m3` to `sensor.octopus_energy_gas_xxx_xxx_previous_accumulative_consumption`

**Reply by gcoan** [reply]
Deleting an entity definitely doesn’t delete the history.  Maybe it gets purged after a period of time but I’ve successfully deleted and recreated entities that retain their history

**Comment by andyblac** [comment]
hmm, did not work for me, I get no history in Energy page after changing it to the new sensor name (this is updating from v9 to v10.0.1)

<img width="1502" alt="Screenshot 2024-01-07 at 10 16 03" src="https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/77936/1e769720-6aa3-47e9-a810-f92de29b4add">

also no history in the entity
<img width="1547" alt="Screenshot 2024-01-07 at 10 15 22" src="https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/77936/c1e1c605-b1a7-4a0e-b0b3-630063c9517d">

**Reply by granthgh** [reply]
I have the same issue with the energy page with 10.0.0.2, so downgraded back to 9.2.1.

**Reply by BottlecapDave** [reply]
Have you followed the process @alfwro13 outlined [here](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/discussions/678#discussioncomment-8064083) to rename your entities?

**Reply by granthgh** [reply]
Sorry,  I hadn't followed the delete/rename. Done that now and it seems to work.
Only have one issue now with the _previous_accumulative_consumption_m3 entity giving an unknown value. However its not that important.

**Reply by BottlecapDave** [reply]
That should populate once yesterday's rates come in.

**Reply by andyblac** [reply]
I had same issue, I had to do the same with `sensor.octopus_energy_gas_xxx_xxx_previous_accumulative_consumption_m3`, delete,  `sensor.octopus_energy_gas_xxxx_xxx_previous_accumulative_consumption` and rename `sensor.octopus_energy_gas_xxx_xxx_previous_accumulative_consumption_m3` to `sensor.octopus_energy_gas_xxx_xxx_previous_accumulative_consumption`


take backup before just incase it don't work.

**Comment by jtetzlaff** [comment]
would those changes affect / cause some sensors to dissapear?
After updating tot he latest version, i've got a number of sensors unavailable;
Electricity Current Rate, Previous Rate, Next rate, Off Peak

And the following log message:
`This error originated from a custom integration.

Logger: custom_components.octopus_energy.coordinators.electricity_rates
Source: helpers/update_coordinator.py:300
Integration: Octopus Energy (documentation, issues)
First occurred: 16:24:19 (17 occurrences)
Last logged: 16:40:19

Unexpected error fetching ELECTRICITY_RATES_101zzzzzzzzz_zzzzzzzzzz data: 'NoneType' object has no attribute 'completed'
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 300, in _async_refresh
    self.data = await self._async_update_data()
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 256, in _async_update_data
    return await self.update_method()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/octopus_energy/coordinators/electricity_rates.py", line 133, in async_update_electricity_rates_data
    dispatches.dispatches.completed
AttributeError: 'NoneType' object has no attribute 'completed'
`

**Reply by BottlecapDave** [reply]
Hi. This should be resolved in the latest version (v10.0.3)
